File: api/formats.js
import { execFile } from "child_process";
import path from "path";

export default async function handler(req, res) {
  const { url } = req.query;
  if (!url) {
    return res.status(400).json({ error: "url is required" });
  }

  const ytdlpPath = path.join(process.cwd(), "bin", "yt-dlp");

  execFile(
    ytdlpPath,
    ["-J", url],
    { maxBuffer: 10 * 1024 * 1024 },
    (err, stdout, stderr) => {
      if (err) {
        return res.status(500).json({
          error: "yt-dlp failed",
          details: stderr,
        });
      }

      const data = JSON.parse(stdout);

      const formats = data.formats.map(f => ({
        format_id: f.format_id,
        ext: f.ext,
        resolution: f.resolution,
        fps: f.fps,
        vcodec: f.vcodec,
        acodec: f.acodec,
        filesize: f.filesize,
        url: f.url, // ← googlevideo.com/videoplayback
      }));

      res.status(200).json({
        title: data.title,
        duration: data.duration,
        formats,
      });
    }
  );
}

File: api/formats.py
import json
import subprocess
from urllib.parse import parse_qs, urlparse

def handler(request):
    url = request.args.get("url")
    if not url:
        return {
            "statusCode": 400,
            "body": json.dumps({"error": "url parameter is required"})
        }

    try:
        # yt-dlp JSON取得
        proc = subprocess.run(
            ["yt-dlp", "-J", url],
            capture_output=True,
            text=True,
            timeout=8
        )

        if proc.returncode != 0:
            return {
                "statusCode": 500,
                "body": json.dumps({"error": proc.stderr})
            }

        data = json.loads(proc.stdout)

        formats = []
        for f in data.get("formats", []):
            if "url" in f:
                formats.append({
                    "format_id": f.get("format_id"),
                    "ext": f.get("ext"),
                    "vcodec": f.get("vcodec"),
                    "acodec": f.get("acodec"),
                    "filesize": f.get("filesize"),
                    "fps": f.get("fps"),
                    "resolution": f.get("resolution"),
                    "url": f["url"],
                })
